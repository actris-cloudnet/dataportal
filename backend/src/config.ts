import {ClientConfiguration} from 'aws-sdk/clients/s3'

interface Config {
  port: number
  connectionName: string
  downloadBaseUrl: string
  pidServiceUrl: string
  pidServiceTimeoutMs: number
  publicDir: string
  s3: s3Config,
  allowUpdateLimitDays: number
  storageService: {
    host: string
    port: string
    user: string
    password: string
  }
}

interface s3Config {
  connection: {
    rw: ClientConfiguration
  },
  buckets: {
    upload: string
  }
}

const testConfig = {
  port: 3001,
  connectionName: 'test',
  downloadBaseUrl: 'http://localhost:3001/api/download/',
  pidServiceUrl: 'http://localhost:5801/pid/',
  pidServiceTimeoutMs: 200,
  publicDir: 'tests/data/public',
  s3: {
    connection: {rw: {}},
    buckets: {upload: 'cloudnet-upload-test'}
  },
  allowUpdateLimitDays: 2,
  storageService: {
    host: 'localhost',
    port: '5910',
    user: 'test',
    password: 'test'
  }
}

const devConfig = {
  port: 3000,
  connectionName: 'default',
  downloadBaseUrl: 'http://localhost:3000/api/download/',
  pidServiceUrl: 'http://localhost:5800/pid/',
  pidServiceTimeoutMs: 2000,
  publicDir: 'public',
  s3: {
    connection: {rw: {}},
    buckets: {upload: 'cloudnet-upload-dev'}
  },
  allowUpdateLimitDays: 2,
  storageService: {
    host: 'localhost',
    port: '5900',
    user: 'test',
    password: 'test'
  }
}

const ciConfig = {
  ...testConfig,
  ...{ connectionName: 'ci-test' }
}

let config: Config

switch (process.env.NODE_ENV) {
case 'production':
  try {
    config = require('./config.prod').default
  } catch (ex) {
    throw new Error('FATAL: Production configuration not found.')
  }
  break
case 'test':
  config = (process.env.CI) ? ciConfig : testConfig
  break
default:
  config = devConfig
  break
}

export default config
