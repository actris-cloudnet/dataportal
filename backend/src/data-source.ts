import "reflect-metadata";
import { DataSource } from "typeorm";
import env from "./lib/env";

export const AppDataSource = new DataSource({
  type: "postgres",
  host: env.TYPEORM_HOST,
  username: env.TYPEORM_USERNAME,
  password: env.TYPEORM_PASSWORD,
  database: env.TYPEORM_DATABASE,
  port: env.TYPEORM_PORT,
  synchronize: env.TYPEORM_SYNCHRONIZE,
  migrations: [env.TYPEORM_MIGRATIONS],
  migrationsRun: env.TYPEORM_MIGRATIONS_RUN,
  logging: env.TYPEORM_LOGGING,
  entities: [env.TYPEORM_ENTITIES],
});
