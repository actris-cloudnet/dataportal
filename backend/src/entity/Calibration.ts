import { Column, Entity, BeforeInsert, BeforeUpdate } from "typeorm";

@Entity()
export class Calibration {
  @Column({ primary: true })
  instrumentPid!: string;

  @Column({ type: "date", primary: true })
  measurementDate!: string;

  @Column({ type: "jsonb" })
  data!: any;

  @Column()
  createdAt!: Date;

  @Column()
  updatedAt!: Date;

  @BeforeInsert()
  updateDateCreation() {
    this.createdAt = new Date();
    this.updatedAt = this.createdAt;
  }

  @BeforeUpdate()
  updateDateUpdate() {
    this.updatedAt = new Date();
  }
}
