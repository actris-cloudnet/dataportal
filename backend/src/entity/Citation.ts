import {Column, Entity, ManyToMany, ManyToOne, PrimaryColumn, PrimaryGeneratedColumn} from 'typeorm'
import {Site} from './Site'
import {Model} from './Model'


@Entity()
export class Citation {

  @PrimaryColumn()
  id!: string

  @Column({type: 'text'})
  acknowledgements!: string

}

@Entity()
export class RegularCitation extends Citation {

  @ManyToMany(_ => Site)
  sites!: Site[]
}

@Entity()
export class ModelCitation extends Citation {

  @ManyToMany(_ => Model)
  models!: Model[]
}
