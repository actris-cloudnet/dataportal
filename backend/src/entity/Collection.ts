import {
  BeforeInsert,
  BeforeUpdate,
  Column,
  Entity,
  JoinTable,
  ManyToMany,
  PrimaryColumn,
} from 'typeorm'
import {File} from './File'
import { v4 as generateUuidV4 } from 'uuid'

@Entity()
export class Collection {

  @PrimaryColumn('uuid')
  uuid!: string

  @ManyToMany(_type => File)
  @JoinTable()
  files!: File[]

  @Column({default: ''})
  title!: string

  @Column({default: ''})
  pid!: string

  @Column({default: 0})
  downloadCount!: number

  @Column()
  createdAt!: Date

  @Column()
  updatedAt!: Date

  @BeforeInsert()
  setCreatedAt() {
    this.createdAt = new Date()
    this.updatedAt = this.createdAt
  }

  @BeforeUpdate()
  setUpdatedAt() {
    this.updatedAt = new Date()
  }

  constructor(files: File[]) {
    this.uuid = generateUuidV4()
    this.files = files
  }
}
