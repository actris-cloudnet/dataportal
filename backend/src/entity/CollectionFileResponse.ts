import {File, ModelFile, RegularFile} from './File'

export class CollectionFileResponse {

  uuid: string
  measurementDate: Date
  site: string
  siteId: string
  product: string
  productId: string
  modelId: string | null
  size: number
  volatile: boolean
  legacy: boolean

  constructor(file: RegularFile | ModelFile) {
    this.uuid = file.uuid
    this.measurementDate = file.measurementDate
    this.site = file.site.humanReadableName
    this.siteId = file.site.id
    this.product = file.product.humanReadableName
    this.productId = file.product.id
    this.modelId = 'model' in file ? file.model.id : null
    this.size = file.size
    this.volatile = file.volatile
    this.legacy = file.legacy
  }
}
