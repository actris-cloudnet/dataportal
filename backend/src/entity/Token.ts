import { Column, Entity, PrimaryColumn, ManyToOne } from "typeorm";
import { timingSafeEqual, createHash } from "node:crypto";
import { UserAccount } from "./UserAccount";

@Entity()
export class Token {
  @PrimaryColumn("bytea")
  selector!: Buffer;

  @Column("bytea")
  verifierHash!: Buffer;

  @ManyToOne(() => UserAccount, { nullable: false })
  userAccount!: UserAccount;

  @Column()
  expiresAt!: Date;

  compareVerifier(verifier: Buffer) {
    return timingSafeEqual(this.verifierHash, hashVerifier(verifier));
  }
}

export function hashVerifier(verifier: Buffer) {
  return createHash("sha256").update(verifier).digest();
}
