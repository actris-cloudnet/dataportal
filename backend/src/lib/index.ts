export const stringify = (obj: any): string => JSON.stringify(obj, null, 2)
