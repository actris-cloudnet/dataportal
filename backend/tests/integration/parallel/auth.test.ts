import axios from "axios";
import { backendPublicUrl } from "../../lib";
import { hashVerifier, Token } from "../../../src/entity/Token";
import { afterAll, beforeAll, describe, expect, it } from "@jest/globals";
import { randomBytes } from "node:crypto";
import { DataSource, Repository } from "typeorm";
import { AppDataSource } from "../../../src/data-source";

const loginUrl = `${backendPublicUrl}auth/login`;
const logoutUrl = `${backendPublicUrl}auth/logout`;
const meUrl = `${backendPublicUrl}auth/me`;
const headers = { "content-type": "application/json" };

let dataSource: DataSource;
let tokenRepo: Repository<Token>;

beforeAll(async () => {
  dataSource = await AppDataSource.initialize();
  tokenRepo = dataSource.getRepository(Token);
});

afterAll(async () => {
  await dataSource.destroy();
});

describe("GET /auth/login", () => {
  it("returns unauthorized error", async () => {
    await expect(axios.post(loginUrl, { username: "hax0r", password: "hax0r" }, { headers })).rejects.toMatchObject({
      response: { status: 401 },
    });
  });

  it("rejects login by submitter", async () => {
    await expect(
      axios.post(loginUrl, { username: "submitter", password: "submitter" }, { headers }),
    ).rejects.toMatchObject({
      response: { status: 401 },
    });
  });

  it("returns admin permissions", async () => {
    const res = await axios.post(loginUrl, { username: "admin", password: "admin" }, { headers });
    expect(res.status).toEqual(200);
    expect(res.data).toMatchSnapshot();
    expect(res.headers).toHaveProperty("set-cookie");
  });
});

describe("GET /auth/me", () => {
  it("returns unauthorized error without cookie", async () => {
    await expect(axios.get(meUrl)).rejects.toMatchObject({ response: { status: 401 } });
  });

  it("returns unauthorized error with invalid cookie", async () => {
    await expect(axios.get(meUrl, { headers: { cookie: "token=abc123" } })).rejects.toMatchObject({
      response: { status: 401 },
    });
  });

  it("returns current user with cookie from login", async () => {
    const resLogin = await axios.post(loginUrl, { username: "admin", password: "admin" }, { headers });
    expect(resLogin.status).toEqual(200);
    expect(resLogin.headers).toHaveProperty("set-cookie");
    const cookie = resLogin.headers["set-cookie"]!.map((c) => c.split(";")[0]);
    const resMe = await axios.get(meUrl, { headers: { cookie } });
    expect(resMe.status).toEqual(200);
    expect(resMe.data).toMatchSnapshot();
  });

  it("returns current user with new cookie", async () => {
    const expiresAt = new Date(Date.now() + 24 * 60 * 60 * 1000);
    const token = await createToken(expiresAt);
    const resMe = await axios.get(meUrl, { headers: { cookie: `token=${token}` } });
    expect(resMe.status).toEqual(200);
    expect(resMe.data).toMatchSnapshot();
  });

  it("returns unauthorized error with expired cookie", async () => {
    const expiresAt = new Date(Date.now() - 24 * 60 * 60 * 1000);
    const token = await createToken(expiresAt);
    await expect(axios.get(meUrl, { headers: { cookie: `token=${token}` } })).rejects.toMatchObject({
      response: { status: 401 },
    });
  });
});

describe("GET /auth/logout", () => {
  it("logs out current user", async () => {
    const resLogin = await axios.post(loginUrl, { username: "admin", password: "admin" }, { headers });
    expect(resLogin.status).toEqual(200);
    const cookie = resLogin.headers["set-cookie"]!.map((c) => c.split(";")[0]);
    const resMe = await axios.get(meUrl, { headers: { cookie } });
    expect(resMe.status).toEqual(200);
    await axios.post(logoutUrl, undefined, { headers: { cookie } });
    expect(resMe.status).toEqual(200);
    await expect(axios.get(meUrl, { headers: { cookie } })).rejects.toMatchObject({ response: { status: 401 } });
  });
});

async function createToken(expiresAt: Date) {
  const selector = randomBytes(16);
  const verifier = randomBytes(16);
  await tokenRepo.insert({
    selector,
    verifierHash: hashVerifier(verifier),
    userAccount: { id: 1 },
    expiresAt,
  });
  return selector.toString("hex") + verifier.toString("hex");
}
