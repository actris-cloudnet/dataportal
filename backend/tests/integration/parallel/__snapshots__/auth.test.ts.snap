// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /auth/login returns admin permissions 1`] = `
{
  "fullName": null,
  "id": 1,
  "orcidId": null,
  "permissions": [
    {
      "id": 1,
      "permission": "canUpload",
    },
    {
      "id": 2,
      "permission": "canUploadModel",
    },
    {
      "id": 3,
      "permission": "canDelete",
    },
    {
      "id": 4,
      "permission": "canProcess",
    },
    {
      "id": 5,
      "permission": "canCalibrate",
    },
    {
      "id": 6,
      "permission": "canGetStats",
    },
    {
      "id": 7,
      "permission": "canAddPublication",
    },
    {
      "id": 8,
      "permission": "canPublishTask",
    },
    {
      "id": 10,
      "permission": "canLogin",
    },
  ],
  "username": "admin",
}
`;

exports[`GET /auth/me returns current user with cookie from login 1`] = `
{
  "fullName": null,
  "id": 1,
  "orcidId": null,
  "permissions": [
    {
      "id": 1,
      "permission": "canUpload",
    },
    {
      "id": 2,
      "permission": "canUploadModel",
    },
    {
      "id": 3,
      "permission": "canDelete",
    },
    {
      "id": 4,
      "permission": "canProcess",
    },
    {
      "id": 5,
      "permission": "canCalibrate",
    },
    {
      "id": 6,
      "permission": "canGetStats",
    },
    {
      "id": 7,
      "permission": "canAddPublication",
    },
    {
      "id": 8,
      "permission": "canPublishTask",
    },
    {
      "id": 10,
      "permission": "canLogin",
    },
  ],
  "username": "admin",
}
`;

exports[`GET /auth/me returns current user with new cookie 1`] = `
{
  "fullName": null,
  "id": 1,
  "orcidId": null,
  "permissions": [
    {
      "id": 1,
      "permission": "canUpload",
    },
    {
      "id": 2,
      "permission": "canUploadModel",
    },
    {
      "id": 3,
      "permission": "canDelete",
    },
    {
      "id": 4,
      "permission": "canProcess",
    },
    {
      "id": 5,
      "permission": "canCalibrate",
    },
    {
      "id": 6,
      "permission": "canGetStats",
    },
    {
      "id": 7,
      "permission": "canAddPublication",
    },
    {
      "id": 8,
      "permission": "canPublishTask",
    },
    {
      "id": 10,
      "permission": "canLogin",
    },
  ],
  "username": "admin",
}
`;
